workflow-trace-viewer/src/jvmMain/kotlin/com/squareup/workflow1/traceviewer/util/parser/DiffUtils.kt

```diff
@@ -0,0 +1,222 @@
+ */
+private fun parseChangedDiff(fullDiff: String): List<Pair<SpanStyle, String>> {
+  val operations: MutableList<Pair<SpanStyle, String>> = mutableListOf()
+  var i = 0
+  while (i < fullDiff.length) {
+    when {
+      fullDiff.startsWith("--", i) -> {
+        val end = fullDiff.indexOf("--", i + 2)
+        if (end != -1) {
+          val removed = fullDiff.substring(i + 2, end)
+          operations.add(DiffStyles.DELETE to removed)
+          i = end + 2
+        }
+      }
+
+      fullDiff.startsWith("++", i) -> {
+        val end = fullDiff.indexOf("++", i + 2)
+        if (end != -1) {
+          val added = fullDiff.substring(i + 2, end)
+          operations.add(DiffStyles.INSERT to added)
+          i = end + 2
+        }
+      }
+
+      else -> {
+        val nextTagStart = listOf(
+          fullDiff.indexOf("--", i),
+          fullDiff.indexOf("++", i)
+        ).filter { it >= 0 }.minOrNull() ?: fullDiff.length
+        operations.add(DiffStyles.UNCHANGED to fullDiff.substring(i, nextTagStart))
+        i = nextTagStart
+      }
+    }
+  }
+
+  return operations
+}
+
+object DiffStyles {
+  val DELETE = SpanStyle(background = Color.Red.copy(alpha = 0.3f))
+  val INSERT = SpanStyle(background = Color.Green.copy(alpha = 0.3f))
+  val NO_CHANGE = SpanStyle(background = Color.LightGray)
+  val UNCHANGED = SpanStyle()
+
+  fun buildStringWithStyle(
+    style: SpanStyle,
+    text: String,
+    builder: AnnotatedString.Builder
+  ) {
+    builder.pushStyle(style)
+    builder.append(text)
+    builder.pop()
+  }
+}
+
+/**
+ * Pull out each "key=value" pair within the field data by looking for a comma. Since plenty of data
+ * include nesting, doing .split or simple regex won't suffice.
+ *
+ * Manually iterates through the fields and changes the depth of the current comma accordingly
+ */
+private fun getFieldsAsList(field: String): List<String> {
+  val fields = mutableListOf<String>()
+  val currentField = StringBuilder()
+  var depth = 0
+  // We skip past the field's Type's name
+  var i = field.indexOf('(') + 1
+
+  while (i < field.length) {
+    val char = field[i]
+    when (char) {
+      '(', '[', '{' -> {
+        depth++
+        currentField.append(char)
+      }
+
+      ')', ']', '}' -> {
+        depth--
+        currentField.append(char)
+      }
+
+      ',' -> {
+        if (depth == 0) { // end of key=value pair
+          fields += currentField.toString().trim()
+          currentField.clear()
+          i++ // skip space in between key value pairs, i.e. "key1=value1, key2=value2"
+        } else { // nested list, so we ignore
+          currentField.append(char)
+        }
+      }
+
+      else -> currentField.append(char)
+    }
+    i++
+  }
+
+  // Just append whatever is left, since there are no trailing commas
+  if (currentField.isNotBlank()) fields += currentField.toString().trim()
+  return fields
+}
+
+private fun extractTypeName(field: String): String {
+  val stateRegex = Regex("""^(\w+)\(""")
```

Comment:
Nit: Pull this out into a constant so it doesn't have to get recompiled every time this function is called.